### DISCUSSION TITLE
How can I check for linearity of a function

### DISCUSSION THREAD

**Discussion by efokken-abb** [casadi/casadi#3382] [Q&A] [answered]
Hi!

I'm writing a python program, that shall get an nlp dictionary ({"x": ..., "g": ..., "f": ...}) and then check, whether it is a linear problem or a nonlinear problem.

If the problem is linear, the program will create a qpsol-type solver, use the highs-backend and accept also integer variables.
If the problem is nonlinear, an nlpsol-type solver is created and in case, integer variables are present, an error will be thrown.

I didn't find a function to check for linearity, so I did the following:

```python
from typing import List

import casadi


def is_linear(x: List[casadi.MX], f_expression: casadi.MX) -> bool:
    z = casadi.vertcat(*[casadi.vec(entry) for entry in x])

    # f = casadi.Function("f", [z], [f_expression])
    Jf = casadi.Function("J", [z], [casadi.jacobian(f_expression, z)])
    Hf = casadi.Function("J", [z], [casadi.jacobian(Jf(z), z)])

    if Hf(z).nnz() > 0:
        return False
    else:
        return True


x = casadi.MX.sym("x")
y = casadi.MX.sym("y")
u = casadi.MX.sym("u")

z = casadi.horzcat(*[x, y])
f = casadi.vertcat(casadi.sin(x) * y + x + y, y**2, u)  # casadi.sin(x) *


print(is_linear([z, u], f))

print(is_linear([z, u], casadi.sum1(z)))
```

This seems to work correctly, for example the above code outputs False and True, as I would expect.
My question is: Do you spot a mistake? I am a bit uneasy with this code, because it seems to me unnatural to first create functions and then extract expressions again from them.
I did try to get it to work with `casadi.jacobian(f_expression, z), but couldn't find the relevant documentation to understand the output of that function. Any pointers on that would also be greatly appreciated.

**Comment by jgillis** [comment, marked as answer]
Haven't looked at your code, but `is_linear` does exist in casadi already. The implementation is based on the more generic `which_depends`.

**Reply by efokken-abb** [reply]
How foolish of me to not search for the exact function name I came up with for my own implementation!
Thanks a lot!

**Reply by jgillis** [reply]
It's even listed in https://web.casadi.org/python-api/ which is like _impressive_ if you keep in mind our bad track record of keeping that documentation page up-to-date..
